From ac3a4cb3c4610b5129ce15564f64142b268d1b1e Mon Sep 17 00:00:00 2001
From: "Wu, Hao" <hao.wu@intel.com>
Date: Sat, 16 May 2015 13:14:07 +0800
Subject: [PATCH] phy-intel-cht: avoid sending usb bus status when vbus is off.

VBUS drop events from EM component may come earlier than the suspend
event (which invoke usb_gadget_vbus_draw 2mA) from USB device stack.
Notify USB_EVENT_ENUMERATED after VBUS drop event, it will cause some
conflict events received by other component (e.g otg-wakelock).
Need to check vbus status before sending USB_EVENT_ENUMERATED 2mA.

Change-Id: I534bbc934723da64d8453b1fd0f4c85effe77355
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-36075
Signed-off-by: Wu, Hao <hao.wu@intel.com>
---
 drivers/usb/phy/phy-intel-cht.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/usb/phy/phy-intel-cht.c b/drivers/usb/phy/phy-intel-cht.c
index a0d58ff84934..8cf94e47326f 100644
--- a/drivers/usb/phy/phy-intel-cht.c
+++ b/drivers/usb/phy/phy-intel-cht.c
@@ -167,6 +167,14 @@ static int cht_otg_set_power(struct usb_phy *phy, unsigned mA)
 	if (!cht_otg_dev)
 		return -ENODEV;
 
+	/* VBUS drop event from EM part may come firstly to phy-intel-cht and
+	 * suspend event from device stack may come later than the VBUS drop
+	 * event. suspend event will trigger usb_gadget_vbus_draw 2mA when
+	 * b_sess_vld is already clear to 0. It will cause conflict events
+	 * received by other components, (e.g otg-wakelock) */
+	if (!cht_otg_dev->fsm.b_sess_vld)
+		return -ENODEV;
+
 	if (phy->state != OTG_STATE_B_PERIPHERAL)
 		dev_err(phy->dev, "ERR: Draw %d mA in state %s\n",
 			mA, usb_otg_state_string(cht_otg_dev->phy.state));
-- 
1.9.1

